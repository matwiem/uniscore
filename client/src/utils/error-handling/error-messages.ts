import { type AxiosError, isAxiosError } from 'axios'

export const errorMessagesMap = {
    UNEXPECTED_ERROR: 'Unexpected error occurred. Please contact Support.',
    CONNECTION_FAILURE: 'Internet connection error. Check your wires!',
    NOT_FOUND: "Broken link. The page you're trying to access doesn't exist.",
    SYSTEM_FAILURE: "Something is wrong, but we're working on it!"
}

export interface UserErrorMessage {
    type: keyof typeof errorMessagesMap
}

export const getErrorMessage = (errMessage: UserErrorMessage) =>
    errorMessagesMap[errMessage.type]

export const HTTPStatusCodeToErrorMessageMap: {
    [code in number]?: keyof typeof errorMessagesMap;
} = {
    404: 'NOT_FOUND'
}

export const getErrorMessageByHTTPStatusCode = (statusCode: number) => {
    return HTTPStatusCodeToErrorMessageMap[statusCode]
}

export const hasErrorCode = (status: number) => (error: any) =>
    isAxiosError(error) && error.response?.status === status

export const isServerError = (e: AxiosError) => {
    if (e.response == null) {
        return false
    }
    const status = e.response.status
    return status >= 500
}

export const isClientError = (e: AxiosError) => {
    if (e.response == null) {
        return false
    }
    const status = e.response.status
    return status >= 400 && status < 500
}

export const isUnauthorizedError = hasErrorCode(401)
export const isForbiddenError = hasErrorCode(403)
export const isInternalServerError = hasErrorCode(403)

export const getUserErrorMessage = (e: unknown): UserErrorMessage => {
    if (!(e instanceof Error)) {
        return { type: 'UNEXPECTED_ERROR' }
    }

    if (e instanceof TypeError) {
        return { type: 'UNEXPECTED_ERROR' }
    }

    if (!isAxiosError(e)) {
        return { type: 'UNEXPECTED_ERROR' }
    }

    if ((e.response == null) || e.code === 'ERR_NETWORK') {
        return { type: 'CONNECTION_FAILURE' }
    }

    const messageType = getErrorMessageIfAxiosError(e)
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (messageType) {
        return { type: messageType }
    }

    if (isServerError(e)) {
        return { type: 'SYSTEM_FAILURE' }
    }

    return { type: 'UNEXPECTED_ERROR' }
}

const getErrorMessageIfAxiosError = (e: AxiosError) => {
    if ((e.response?.status) != null) {
        return getErrorMessageByHTTPStatusCode(e.response.status)
    }
}
