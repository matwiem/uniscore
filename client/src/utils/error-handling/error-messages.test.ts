/* eslint-disable jest/no-jasmine-globals,jest/no-conditional-expect */
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import axios from 'axios'

import { getUserErrorMessage } from './error-messages'

const handlers = [
    rest.get('/mocks/errors/500', (req, res, context) => {
        return res(context.status(500, 'Internal Server Error'))
    }),
    rest.get('/mocks/errors/404', (req, res, context) => {
        return res(context.status(404, 'Not Found'))
    }),
    rest.get('/mocks/errors/network', (req, res, context) => {
        res.networkError('Network Error')
    })
]

const server = setupServer(...handlers)

describe('getUserErrorMessage', () => {
    beforeAll(() => {
        server.listen({
            onUnhandledRequest: 'error'
        })
    })
    beforeEach(() => { server.resetHandlers() })
    afterAll(() => { server.close() })

    it('should produce UNEXPECTED_ERROR if a non-error was thrown', () => {
        expect(getUserErrorMessage(1)).toMatchInlineSnapshot(`
      Object {
        "type": "UNEXPECTED_ERROR",
      }
    `)
    })

    it('should produce UNEXPECTED_ERROR if a TypeError was thrown', () => {
        const error = new TypeError('undefined is not a function')
        expect(getUserErrorMessage(error)).toMatchInlineSnapshot(`
      Object {
        "type": "UNEXPECTED_ERROR",
      }
    `)
    })

    it('should produce CONNECTION_FAILURE if a HTTP NetworkError was thrown', async () => {
        try {
            await axios.get('/mocks/errors/network')
            fail(
                "error should be thrown, if it's not then probably HTTP is misconfigured"
            )
        } catch (error: unknown) {
            expect(getUserErrorMessage(error)).toMatchInlineSnapshot(`
        Object {
          "type": "CONNECTION_FAILURE",
        }
      `)
        }
    })

    const testcases = [
        ['SYSTEM_FAILURE', 500],
        ['NOT_FOUND', 404]
    ]

    it.each(testcases)('should produce %s if a HTTP %s was thrown', async (errorMessage, statusCode) => {
        try {
            await axios.get(`/mocks/errors/${statusCode}`)
            fail("error should be thrown, if it's not then probably HTTP is misconfigured")
        } catch (error: unknown) {
            expect(getUserErrorMessage(error)).toMatchSnapshot()
        }
    })
})
