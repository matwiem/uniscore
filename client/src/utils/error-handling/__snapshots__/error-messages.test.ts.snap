// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUserErrorMessage should produce NOT_FOUND if a HTTP 404 was thrown 1`] = `
Object {
  "type": "NOT_FOUND",
}
`;

exports[`getUserErrorMessage should produce SYSTEM_FAILURE if a HTTP 500 was thrown 1`] = `
Object {
  "type": "SYSTEM_FAILURE",
}
`;
